### ISSUE TITLE
Knative Serving Webhook installation is flaky

### ISSUE THREAD

**Issue by nachtmaar** [kyma-project/kyma#7127]
**Description**

TL;DR
In certain situations the `MutatingWebhookConfiguration` of `Knative Serving` does not get installed, rendering the creation of Knative Services with a custom `ReadinessProbe` unusable.

We traced the error back to this [PR](https://github.com/kyma-project/kyma/pull/6639/files#diff-8e48b2f9f69f051e7bf1d75b04631bedR342). The problem is the templating in [serving.yaml](https://github.com/piotrmsc/kyma/blob/b5fae224cc0b72f404d0ab0ce696bddda5d09b66/resources/knative-serving/templates/serving.yaml#L342), more exactly `{{ if .Release.IsInstall }}`.

From the [Helm Documentation](https://helm.sh/docs/chart_template_guide/builtin_objects/): 
`Release.IsInstall: This is set to true if the current operation is an install.`
`Release.IsUpgrade: This is set to true if the current operation is an upgrade or rollback.`
Because `kyma-operator` retries a failed install with an upgrade, the value is false when installing it a second time.

**Expected result**
Knative Serving mutating webhook get's installed.

**Actual result**
Knative Serving mutating webhook wasn't installed.

```shell
$ k get mutatingwebhookconfigurations.admissionregistration.k8s.io --all-namespaces
NAME                                                 CREATED AT
cluster-essentials-pod-preset-webhook                2020-02-05T14:23:09Z
istio-sidecar-injector                               2020-02-05T14:26:36Z
pod-ready.config.common-webhooks.networking.gke.io   2020-02-05T14:21:08Z
service-catalog-catalog-webhook                      2020-02-05T14:33:52Z
webhook.eventing.knative.dev                         2020-02-05T14:30:09Z
webhook.serving.knative.dev                          2020-02-05T14:28:07Z <- this entry is missing in a cluster where the problem occurs
```

**Problem in Release Cluster**

While testing the Kyma 1.10rc1 cluster, we found out that there was a problem while creating a Knative Service.

If the Knative serving mutating webhook is not installed, `HttpSource`s will be unready

```shell
$ k get httpsources.sources.kyma-project.io --all-namespaces
NAMESPACE          NAME                    READY   REASON
kyma-integration   selenide-jk1-app-name   False   SinkNotFound
kyma-integration   tunas-oneclick          False   SinkNotFound
kyma-integration   tunas-varkes            False   SinkNotFound
```

The problem is that the EventSourcesController can't create a Knative Service.

```shell
$ k logs -n kyma-system event-sources-controller-manager-6756b4ccd-qwhrq controller -f
...
{"level":"error","ts":"2020-02-05T14:00:12.943Z","logger":"event_sources_controller.http-source-controller","caller":"controller/controller.go:357","msg":"Reconcile error","knative.dev/controller":"http-source-controller","error":"failed to create Knative Service: admission webhook \"validation.webhook.serving.knative.dev\" denied the request: validation failed: expected 1 <= 0 <= 2147483647: spec.template.spec.containers[0].readinessProbe.successThreshold\nmissing field(s): spec","errorVerbose":"admission webhook \"validation.webhook.serving.knative.dev\" denied the request: validation failed: expected 1 <= 0 <= 2147483647: spec.template.spec.containers[0].readinessProbe.successThreshold\nmissing field(s): spec\nfailed to create Knative Service\ngithub.com/kyma-project/kyma/components/event-sources/reconciler/httpsource.(*Reconciler).getOrCreateKnService\n\t/workspace/go/src/github.com/kyma-project/kyma/components/event-sources/reconciler/httpsource/reconcile.go:185\ngithub.com/kyma-project/kyma/components/event-sources/reconciler/httpsource.(*Reconciler).reconcileAdapter\n\t/workspace/go/src/github.com/kyma-project/kyma/components/event-sources/reconciler/httpsource/reconcile.go:159\ngithub.com/kyma-project/kyma/components/event-sources/reconciler/httpsource.(*Reconciler).Reconcile\n\t/workspace/go/src/github.com/kyma-project/kyma/components/event-sources/reconciler/httpsource/reconcile.go:102\ngithub.com/kyma-project/kyma/components/event-sources/vendor/knative.dev/pkg/controller.(*Impl).processNextWorkItem\n\t/workspace/go/src/github.com/kyma-project/kyma/components/event-sources/vendor/knative.dev/pkg/controller/controller.go:342\ngithub.com/kyma-project/kyma/components/event-sources/vendor/knative.dev/pkg/controller.(*Impl).Run.func2\n\t/workspace/go/src/github.com/kyma-project/kyma/components/event-sources/vendor/knative.dev/pkg/controller/controller.go:291\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1337","stacktrace":"github.com/kyma-project/kyma/components/event-sources/vendor/knative.dev/pkg/controller.(*Impl).handleErr\n\t/workspace/go/src/github.com/kyma-project/kyma/components/event-sources/vendor/knative.dev/pkg/controller/controller.go:357\ngithub.com/kyma-project/kyma/components/event-sources/vendor/knative.dev/pkg/controller.(*Impl).processNextWorkItem\n\t/workspace/go/src/github.com/kyma-project/kyma/components/event-sources/vendor/knative.dev/pkg/controller/controller.go:343\ngithub.com/kyma-project/kyma/components/event-sources/vendor/knative.dev/pkg/controller.(*Impl).Run.func2\n\t/workspace/go/src/github.com/kyma-project/kyma/components/event-sources/vendor/knative.dev/pkg/controller/controller.go:291"}
{"level":"info","ts":"2020-02-05T14:00:12.944Z","logger":"event_sources_controller.http-source-controller","caller":"controller/controller.go:344","msg":"Reconcile failed. Time taken: 9.491863ms.","knative.dev/controller":"http-source-controller","knative.dev/traceid":"2d1aba86-4f09-438b-a1ac-7a5e72a893aa","knative.dev/key":"kyma-integration/selenide-jk1-app-name"}
```

The value `spec.template.spec.containers[0].readinessProbe.successThreshold` inside the Knative Service is not set by the controller (defaults to 0 if not set). However, the knative serving mutating webhook sets it to 1. The validating webhook of knative serving checks if the value is between `1 <= 0 <= 2147483647`. Because the validation webhook is active but not the defaulter, the value is 0 and the validater complains. Hence rendering knative eventing mesh unusable.

**Steps to reproduce**

Probably these steps (untested):
- install knative-serving via helm and let the installation fail
- trigger a helm update (not install!) => 
- check that there is no serving webhook

**Troubleshooting**

Quickfix:

```shell
# install webhook.serving.knative.dev manually
$ cat << EOF | kubectl apply -f -
apiVersion: admissionregistration.k8s.io/v1beta1
kind: MutatingWebhookConfiguration
metadata:
  labels:
    serving.knative.dev/release: "{{ .Values.global.version }}"
  name: webhook.serving.knative.dev
webhooks:
- admissionReviewVersions:
  - v1beta1
  clientConfig:
    service:
      name: webhook
      namespace: knative-serving
  failurePolicy: Fail
  name: webhook.serving.knative.dev
EOF
```

**Comment by piotrmsc** [comment]
From what I can see here, you can delete the `Release.isInstall` and let it install/upgrade always like other webhooks. You can ask knative guys if it should be changed globally :) This change comes from knative release so I cannot tell you why it has this if, maybe there is a particular reason but like I said, I would suggest asking knative guys. From my POV it can be done like with other webhooks ;-)

**Comment by kubadz** [comment]
It is not from knative release. I had to add these after we bumped knative-serving to version v0.11.1 in the mentioned PR with Istio upgrade. The reason for adding it was that in the previous version of knative-serving `webhook` pod automatically creates a secret, but in newer it doesn't, and secret is placed in chart templates. Same applies for mutatingWebhookConfiguration. Without the `Release.isInstall` helm returns information that those resources already exist, but are not a part of the previous release (what makes sense, as they were created by a job). You can try to use the following helm hooks:
```
    helm.sh/hook: pre-install,pre-upgrade
    helm.sh/hook-delete-policy: before-hook-creation
```

**Comment by michal-hudy** [comment]
Fix has been merged: https://github.com/kyma-project/kyma/pull/7131

**Comment by magicmatatjahu** [comment]
@kubadz Thanks for report! But we needed remove old resources, created internally by KNative Serving release, and hooks suggested by you cannot do it, so only solution was a pre-upgrade job :)
